Node markdown generator tool

SHKnudsen opened the pull request:
### Purpose

This is still very much WIP...

### Declarations

Check these if you believe they are true

- [ ] The codebase is in a better state after this PR
- [ ] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [ ] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.
- [ ] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions) and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.
- [ ] This PR modifies some build requirements and the readme is updated

### Reviewers

### FYIs

SHKnudsen committed:
initial commit

SHKnudsen committed:
Use CLRDLLModule

SHKnudsen committed:
Refactoring

SHKnudsen committed:
add hosts flag and other additions to frompackage

SHKnudsen committed:
🧹🧹

SHKnudsen committed:
Update Program.cs

SHKnudsen committed:
clean up + handle overloads

SHKnudsen committed:
🧹

SHKnudsen committed:
Update AssemblyHandler.cs

SHKnudsen committed:
too many things 🤯

SHKnudsen committed:
Save images instead of embedding

alvpickmans commented on src/DynamoCore/Core/NodeModelAssemblyLoader.cs:
```diff
@@ -199,6 +199,44 @@ internal static bool ContainsNodeModelSubType(Assembly assem)
 
+        internal static bool ContainsNodeModelSubTypeReflectionLoaded(Assembly assem, Type nodeModelType = null)
+        {
+            if(!(nodeModelType is null))
```
This might be easier to read instead of negating a is null assertion
```suggestion
            if(nodeModelType is null)
                return assem.GetTypes().Any(x => IsNodeSubType(x));
            return assem.GetTypes().Any(x => IsNodeSubTypeReflectionLoaded(x, nodeModelType));
            
```

alvpickmans commented on src/DynamoCore/Core/NodeModelAssemblyLoader.cs:
```diff
@@ -199,6 +199,44 @@ internal static bool ContainsNodeModelSubType(Assembly assem)
+            {
+                isNodeSubType = !t.IsAbstract &&
+                    t.IsSubclassOf(nodeModelType) &&
+                    t.GetConstructor(Type.EmptyTypes) != null;
```
From this I read that when trying to get a parameterless ctor when there's none it will throw an exception?
Might be better to use a linq query like `t.GetConstructors().FirstOrDefault(ctor => ctor.GetParameters().Lengh == 0) != null`, as it will reduce memory allocations when throwing an exception

radumg replied:
would that LINQ query above not do more work though @alvpickmans ?
for each constructor it would call `GetParameters()` and only afterwards see if there's one with no params.

way i read @SHKnudsen's line is it tries to get the parameterless ctor directly so fewer reflection calls ?
though not sure i understand what you mean with the memory allocations when throwing ?

alvpickmans replied:
From my understanding, is better to avoid raising an exception if possible as when doing so the exception collects a snapshot of the state of the application (stacktrace and all that stuff).

I don't know how it compares in performance between reflection vs throwing exceptions, but using reflection it explicitly tells what the intent is.

alvpickmans commented on src/DynamoCore/Graph/Nodes/TypeLoadData.cs:
```diff
@@ -97,6 +98,67 @@ public TypeLoadData(Type typeIn)
+
+            ObsoleteMessage = string.Join(
+                "\n",
+                attributes.Where(x=>x is ObsoleteAttribute)
```
```suggestion
                attributes.OfType<ObsoleteAttribute>()
```

alvpickmans commented on src/DynamoCore/Graph/Nodes/TypeLoadData.cs:
```diff
@@ -97,6 +98,67 @@ public TypeLoadData(Type typeIn)
+                    .Cast<ObsoleteAttribute>()
+                    .Select(x => string.IsNullOrEmpty(x.Message) ? "Obsolete" : x.Message));
+
+            IsDeprecated = attributes.Where(x=>x is NodeDeprecatedAttribute).Cast<NodeDeprecatedAttribute>().Any();
```
```suggestion
            IsDeprecated = attributes.OfType<NodeDeprecatedAttribute>().Any();
```
Also if you just need the count, no need to cast them?

SHKnudsen replied:
no....... not sure why i did that 😬 probably copy-pasted this from one where i needed something on the attribute

alvpickmans commented on src/DynamoCore/Graph/Nodes/TypeLoadData.cs:
```diff
@@ -97,6 +98,67 @@ public TypeLoadData(Type typeIn)
+                    .Select(x => string.IsNullOrEmpty(x.Message) ? "Obsolete" : x.Message));
+
+            IsDeprecated = attributes.Where(x=>x is NodeDeprecatedAttribute).Cast<NodeDeprecatedAttribute>().Any();
+            IsMetaNode = attributes.Where(x => x is IsMetaNodeAttribute).Cast<IsMetaNodeAttribute>().Any();
```
Same suggestion as above, use `OfType()`

alvpickmans commented on src/DynamoCore/Graph/Nodes/TypeLoadData.cs:
```diff
@@ -97,6 +98,67 @@ public TypeLoadData(Type typeIn)
+
+            IsDeprecated = attributes.Where(x=>x is NodeDeprecatedAttribute).Cast<NodeDeprecatedAttribute>().Any();
+            IsMetaNode = attributes.Where(x => x is IsMetaNodeAttribute).Cast<IsMetaNodeAttribute>().Any();
+            IsDSCompatible = attributes.Where(x => x is IsDesignScriptCompatibleAttribute).Cast<IsDesignScriptCompatibleAttribute>().Any();
```
Same suggestion as above, use `OfType()`

alvpickmans commented on src/DynamoCore/Graph/Nodes/TypeLoadData.cs:
```diff
@@ -97,6 +98,67 @@ public TypeLoadData(Type typeIn)
+            IsDeprecated = attributes.Where(x=>x is NodeDeprecatedAttribute).Cast<NodeDeprecatedAttribute>().Any();
+            IsMetaNode = attributes.Where(x => x is IsMetaNodeAttribute).Cast<IsMetaNodeAttribute>().Any();
+            IsDSCompatible = attributes.Where(x => x is IsDesignScriptCompatibleAttribute).Cast<IsDesignScriptCompatibleAttribute>().Any();
+            IsHidden = attributes.Where(x=>x is IsVisibleInDynamoLibraryAttribute).Cast<IsVisibleInDynamoLibraryAttribute>().Any(attr => !attr.Visible);
```
Same suggestion as above, use `OfType()`

alvpickmans commented on src/DynamoCore/Graph/Nodes/TypeLoadData.cs:
```diff
@@ -97,6 +98,67 @@ public TypeLoadData(Type typeIn)
+            IsDSCompatible = attributes.Where(x => x is IsDesignScriptCompatibleAttribute).Cast<IsDesignScriptCompatibleAttribute>().Any();
+            IsHidden = attributes.Where(x=>x is IsVisibleInDynamoLibraryAttribute).Cast<IsVisibleInDynamoLibraryAttribute>().Any(attr => !attr.Visible);
+
+            var attribs = attributes.Where(x=>x is NodeNameAttribute).Cast<NodeNameAttribute>();
```
Same suggestion as above, use `OfType()`

alvpickmans commented on src/DynamoCore/Library/FunctionDescriptor.cs:
```diff
@@ -568,4 +569,217 @@ private string GetRootCategory()
+        }
+    }
+
+    //internal static class FunctionDescriptorUtils
```
Remove commented code

alvpickmans commented on src/DynamoCore/Library/FunctionDescriptor.cs:
```diff
@@ -568,4 +569,217 @@ private string GetRootCategory()
+        private readonly LibraryCustomization libraryCustomization;
+        private readonly Assembly reflectionAsm;
+
+        public new string Category
```
Any chance `FunctionDescriptor` can be flaged as `virtual`? That way here you could do `public override string Category` instead. 
Only issue with the `new string Category` is when casting the `ReflectionFunctionDescriptor as FunctionDescriptor` in that case `FunctionDescriptor.Category` will be used instead.

alvpickmans replied:
Also, if the intent is to defer the `Category` computation to when is actually needed, consider using `Lazy<string> category` so it only executes the `get` code once.
```csharp
    private string Lazy<string> category = new Lazy<string>(() => {
        // Code for getting the category goes here  
    });
   
    public override string Category => category.Value;
```

SHKnudsen replied:
> Only issue with the new string Category is when casting the ReflectionFunctionDescriptor as FunctionDescriptor in that case FunctionDescriptor.Category will be used instead.

Yea thats actually a pretty good point! 
Not completely sure if that would be considered an API-breaking change though. Ill bring it up with the dynamo team.

https://docs.microsoft.com/en-us/dotnet/core/compatibility/
![image](https://user-images.githubusercontent.com/13732445/119478976-1d315780-bd48-11eb-8598-91f77e06c482.png)

SHKnudsen replied:
Also, the need for this ReflectionFunctionDescriptor is simply that the normal `FunctionDescriptor` uses `GetCustomAttribute` which is not possible when types are loaded in reflection only, so the `ReflectionFunctionDescriptor` simply provides a different implementation that can be used with reflection only types

alvpickmans commented on src/DynamoCore/Library/FunctionDescriptor.cs:
```diff
@@ -568,4 +569,217 @@ private string GetRootCategory()
+                                .Where(x=>x.FullName == ClassName)
+                                .FirstOrDefault();
+
+                            //get NodeCategoryAttribute for this function if it was been defined
```
Remove commented code?

alvpickmans commented on src/DynamoCore/Library/FunctionDescriptor.cs:
```diff
@@ -568,4 +569,217 @@ private string GetRootCategory()
+                if (ClassName != null)
+                {
+
+                    if (reflectionAsm != null && reflectionAsm.GetType(ClassName) != null)
```
```suggestion
                    if (reflectionAsm != null && reflectionAsm.GetType(ClassName) is Type type)
```
This should give you the required type in an atomic operation and remove the need for the `where` query below

alvpickmans commented on src/DynamoCore/Library/FunctionDescriptor.cs:
```diff
@@ -568,4 +569,217 @@ private string GetRootCategory()
+                            var nodeCat = string.Empty;
+                            foreach (var t in type.DeclaredMembers)
+                            {
+                                if (t.Name == FunctionName)
```
Think about bailing early on these nested blocks to improve readability
```csharp
foreach(var t in type.DeclaredMembers)
{
    if(t.Name != FunctionName)
        continue;
    
    // Code for t.Name == FunctioName
}

alvpickmans commented on src/DynamoCore/Search/SearchElements/ZeroTouchSearchElement.cs:
```diff
@@ -136,4 +136,51 @@ private string GetIconName()
+        private readonly string fullname;
+        public override string CreationName { get { return functionDescriptor != null ? functionDescriptor.MangledName : this.Name; } }
+
+        public override string FullName { get { return fullname; } }
```
```suggestion
        public override string FullName => fullname;
```
I prefer the arrow getter, but independently of which option should be consistent with the rest of the code (`Descriptor` uses arrow while this `get` and `return`)

alvpickmans commented on src/DynamoCore/Search/SearchElements/ZeroTouchSearchElement.cs:
```diff
@@ -136,4 +136,51 @@ private string GetIconName()
+
+        protected override NodeModel ConstructNewNodeModel()
+        {
+            throw new NotImplementedException();
```
If this doesn have an implementation might be better to to `base.ConstructNewNodeModel()` or even not overriding the method?

SHKnudsen replied:
the ConstructNewNodeModel is marked abstract in the base, so i have to implement it here. However, as this is meant to be used in ReflectionOnly we cant create anything from these types as they are only loaded into the `MetaDataLoadContext`.

alvpickmans commented on src/Engine/ProtoCore/Reflection/MemberInfoReflectionExtensions.cs:
```diff
@@ -0,0 +1,211 @@
+                }
+
+                // This will fail if there are overloads of the method
+                var baseTypeMethod = baseType.GetMethod(methodInfo.Name);
```
If the expected behaviour with overloads is to return `null`, would it be better to use a `Where` filter query and check the methods with that name is 1.

If > 1, return null.

alvpickmans commented on src/Engine/ProtoCore/Reflection/MemberInfoReflectionExtensions.cs:
```diff
@@ -0,0 +1,211 @@
+                }
+
+                // Need to return the original methodInfo if the method is hiding the inherited member using the new keyword
+                // This is probably not a supported way of doing this... but getting desperate
```
Love this comment. Although no idea what the edge case is ha

alvpickmans commented on src/Engine/ProtoCore/Reflection/MemberInfoReflectionExtensions.cs:
```diff
@@ -0,0 +1,211 @@
+
+        private static Attribute CreateInstance(System.Type type, object args)
+        {
+            var loadedAssembly = Assembly.LoadFrom(type.Assembly.Location);
+            // Check if in right assembly before laoding...
+            var loadedType = loadedAssembly.GetType(type.FullName);
```
Might no be understanding this lines, but why cannot this use the input `Type` directly instead of loading its assembly (which should be loaded already as you have its type as input?) to get the type again?

SHKnudsen replied:
the type will in this case only be loaded into reflection context (or the MetadataLoadContext as im using for this), when this is the case you cannot create instances from it, so i need to actually load it before its possible to create instances from the type. We are only using this to get Dynamo defined attributes so we know that the assemblies that defines these types are available to us and there for we can load the assembly and the get the type from the Loaded assembly instead of the MLC assembly

alvpickmans replied:
This will force the assembly to be loaded then. Don't have much context to say but if this is the desired behaviour then it's fine. Been recently pulling my hair dealing with assemblies and app domains and is not pretty

alvpickmans commented on src/NodeDocumentationMarkdownGenerator/MarkdownHandler.cs:
```diff
@@ -0,0 +1,288 @@
+{
+    internal static class MarkdownHandler
+    {
+        internal static void CreateMdFilesFromFileNames(List<MdFileInfo> fileInfos, string outputDir, bool overWrite, ILogger logger, bool compressImages = false, string dictionaryPath = null, string layoutSpec = null)
```
Helpful to add xml documentation here

alvpickmans commented on src/NodeDocumentationMarkdownGenerator/MdFileInfo.cs:
```diff
@@ -0,0 +1,96 @@
+
+        public MdFileInfo(string nodeName, string nodeNamespace, string fullCategory, string externalLib, string qualifiedName)
+        {
+            NodeName = nodeName ?? throw new ArgumentNullException(nameof(nodeName));
+            NodeNamespace = nodeNamespace ?? throw new ArgumentNullException(nameof(nodeNamespace));
```
Maybe check if they are empty as well?

SHKnudsen committed:
comment updates

radumg commented on src/DynamoCore/Core/NodeModelAssemblyLoader.cs:
```diff
@@ -199,6 +199,45 @@ internal static bool ContainsNodeModelSubType(Assembly assem)
+            return assem.GetTypes().Any(x => IsNodeSubTypeReflectionLoaded(x, nodeModelType));
+        }
+
+        internal static bool IsNodeSubTypeReflectionLoaded(Type t, Type nodeModelType)
```
can `nodeModelType` be different types ? or will it always be the `NodeModel` type from Dynamo core ?
if the latter, you can use the constant expression `typeof(NodeModel)` instead of exposing as parameter.

SHKnudsen replied:
Yea so the issue with that is i need to make sure that its using a reference to the same type. because im calling this method with a type coming form the MLC context, i need to also compare it with a type from the MLC. Doing `typeof(NodeModel)` will get me the NodeModel type from the normal context (not sure what the right definition is of that 🤡) and thoese two dont point to the same refernce, so it will always return false.

Another option is to compare the name, but dont think that's as safe as this, what do you think?

radumg replied:
ah, gotcha, makes sense, didn't know the same type would not equal itself if it's from different contexts (MLC vs normal)

radumg commented on src/DynamoCore/Core/NodeModelAssemblyLoader.cs:
```diff
@@ -199,6 +199,45 @@ internal static bool ContainsNodeModelSubType(Assembly assem)
+
+                isNodeSubType = customAttributes.
+                    Any(x => 
+                    x.AttributeType.Name.Equals(nameof(NodeNameAttribute)) ||
```
should probs put those eligible names in a `private const` list/array/hashset on this class and just check if it contains the name

`.Any(x => list.Contains(x.AttributeType.Name))`

radumg commented on src/DynamoCore/Graph/Nodes/TypeLoadData.cs:
```diff
@@ -97,6 +98,68 @@ public TypeLoadData(Type typeIn)
+
+            ObsoleteMessage = string.Join(
+                "\n",
+                attributes.Where(x=>x is ObsoleteAttribute)
```
still have a Where/Cast here -> OfType

radumg commented on src/DynamoCore/Graph/Nodes/TypeLoadData.cs:
```diff
@@ -97,6 +98,68 @@ public TypeLoadData(Type typeIn)
                 .SelectMany(x => x.PortTypes);
         }
 
+        public TypeLoadData (Type typeIn, ICollection<Attribute> attributes)
```
XML comment on public ctor

radumg commented on src/DynamoCore/Library/FunctionDescriptor.cs:
```diff
@@ -568,4 +569,117 @@ private string GetRootCategory()
+                categoryBuf.Append(GetRootCategory());
+
+                //if this is not BuiltIn function search NodeCategoryAttribute for it
+                if (ClassName != null)
```
this block should probs be a private method as it's quite long

radumg commented on src/DynamoCore/Search/SearchElements/NodeModelSearchElement.cs:
```diff
@@ -26,6 +26,12 @@ internal NodeModelSearchElement(TypeLoadData typeLoadData) : base(typeLoadData)
 
+        internal NodeModelSearchElement(TypeLoadData typeLoadData, Type baseType) : base(typeLoadData)
+        {
+            //constructor = typeLoadData.Type.GetDefaultConstructor < baseType.GetType() > (;
```
delete commented line

radumg commented on src/DynamoCoreWpf/ViewModels/Core/DynamoViewModelEventArgs.cs:
```diff
@@ -183,6 +183,12 @@ public class OpenNodeAnnotationEventArgs : OpenDocumentationLinkEventArgs
+        /// <summary>
+        /// Name of the package this node belongs to
+        /// </summary>
+        public string PkgName => Category.Split('.').FirstOrDefault();
```
suggest using full name `PackageName` for prop

radumg commented on src/Engine/ProtoCore/FFI/CLRDLLModule.cs:
```diff
@@ -179,6 +180,11 @@ public static MethodInfo DisposeMethod
         /// </summary>
         public CLRDLLModule Module { get; private set; }
 
+        public bool ReflectionContext
```
suggest naming props that return boo with `Is`, so `IsReflectionContext` - much more descriptive and easy to understand what it does

radumg replied:
can also use `=>` if that's appropriate in this codebase ?

radumg commented on src/Engine/ProtoCore/FFI/CLRDLLModule.cs:
```diff
@@ -393,18 +403,32 @@ private ClassDeclNode ParseSystemType(Type type, string alias)
             
             foreach (var m in methods)
             {
+                var na = m.Name;
```
use full name for vars, especially in context of low-level layer that's tricky to understand :D

radumg commented on src/Engine/ProtoCore/FFI/CLRDLLModule.cs:
```diff
@@ -393,18 +403,32 @@ private ClassDeclNode ParseSystemType(Type type, string alias)
+                if (m.DeclaringType.Assembly.ReflectionOnly)
+                {
+                    baseDefinition = m.GetBaseDefinitionReflectionContext();
+                    if (baseDefinition is null)
+                        continue;
```
could we do this check after the `else` ?
ie. would `m.GetBaseDefinition()` ever return `null` ? if so, then would be good to move it so we check for null in both these cases

SHKnudsen replied:
yea so the main if statement here uses `GetBaseDefinitionReflectionContext` which can potentially return null as there are cases where some of the checks that it performs fails because of the reflection context. In the `else` statement we are using `GetBaseDefinition` which will never return null, it will either return the base definition or the same methodInfo. Which is why im only checking for null here.

radumg commented on src/Engine/ProtoCore/FFI/CLRDLLModule.cs:
```diff
@@ -538,10 +564,18 @@ private bool isOverloadedOperator(MethodInfo m)
+                ParameterInfo[] ps = m.GetParameters();
+                if ((ps == null || ps.Length == 0) && m.Name == "Dispose")
+                    return true;
+                return false;
```
could be simplified to `return (condition)` so no need for the `if`

radumg commented on src/Engine/ProtoCore/FFI/CLRDLLModule.cs:
```diff
@@ -585,10 +629,19 @@ private ProtoCore.AST.AssociativeAST.AssociativeNode ParseProperty(PropertyInfo
+                    varDeclNode.IsStatic = m.IsStatic;
+                return varDeclNode;
+            }
+            catch (Exception)
```
remove `Exception` if we're not using it in `catch`

radumg commented on src/Engine/ProtoCore/FFI/CLRDLLModule.cs:
```diff
@@ -601,18 +654,33 @@ private static bool SupressesImport(MemberInfo member, Dictionary<MethodInfo, At
                 return true;
 
-            object[] atts = member.GetCustomAttributes(false);
+            var reflectionContext = member.Module.Assembly.ReflectionOnly;
```
same suggestion as property, name it with `isReflectionContext`, makes it easier to read code

radumg commented on src/Engine/ProtoCore/FFI/CLRDLLModule.cs:
```diff
@@ -601,18 +654,33 @@ private static bool SupressesImport(MemberInfo member, Dictionary<MethodInfo, At
+                            return true;
+                    }
+                }
+                catch (Exception)
```
same as above, empty catches don't need the exception

radumg commented on src/Engine/ProtoCore/FFI/CLRDLLModule.cs:
```diff
@@ -601,18 +654,33 @@ private static bool SupressesImport(MemberInfo member, Dictionary<MethodInfo, At
+            }
 
             var method = member as MethodInfo;
             if (method != null)
```
could we bail early here instead of a rather large `if` block ?

radumg commented on src/Engine/ProtoCore/FFI/CLRDLLModule.cs:
```diff
@@ -651,7 +720,27 @@ private ProtoCore.AST.AssociativeAST.VarDeclNode ParseFieldDeclaration(FieldInfo
+            {
+                varDeclNode = ParseArgumentDeclaration(f.Name, f.FieldType);
+            }
+
```
no space between `if`/`else` blocks

radumg commented on src/Engine/ProtoCore/FFI/CLRDLLModule.cs:
```diff
@@ -665,11 +754,28 @@ private FunctionDefinitionNode ParseFieldAccessor(FieldInfo f)
+            ProtoCore.Type returnType = new ProtoCore.Type();
+            if (!ReflectionContext)
+                returnType = CLRModuleType.GetProtoCoreType(f.FieldType, Module);
+
```
no break here either

radumg commented on src/Engine/ProtoCore/FFI/CLRDLLModule.cs:
```diff
@@ -665,11 +754,28 @@ private FunctionDefinitionNode ParseFieldAccessor(FieldInfo f)
 
+            ProtoCore.Type returnType = new ProtoCore.Type();
+            if (!ReflectionContext)
+                returnType = CLRModuleType.GetProtoCoreType(f.FieldType, Module);
```
all other `if` blocks have `{}`

radumg commented on src/Engine/ProtoCore/FFI/CLRDLLModule.cs:
```diff
@@ -845,8 +966,13 @@ private ArgumentSignatureNode ParseArgumentSignature(MethodBase method)
+
+            argumentSignature.IsVarArg = parameters.Any() && 
+                (ReflectionContext ? 
+                    parameters.Last().AttributesFromReflectionContext().Any(x => x is ParamArrayAttribute) : 
```
suggest doing `.OfType<ParamArrayAttribute>().Any()` instead

radumg replied:
will match line below it better too

radumg commented on src/Engine/ProtoCore/FFI/CLRDLLModule.cs:
```diff
@@ -988,6 +1118,27 @@ public override FFIFunctionPointer GetFunctionPointer(string className, string n
+            {
+                var tn = type.FullName;
+                //For now there is no support for generic type.
+                if (!type.IsGenericType && type.IsPublic && !CLRModuleType.SupressesImport(type))
```
do a bail early (`continue`) here instead of `if` block

radumg commented on src/Engine/ProtoCore/FFI/CLRDLLModule.cs:
```diff
@@ -1141,6 +1292,22 @@ private Type[] GetTypes(string typeName)
+            {
+                if (Module is null)
+                    types = Assembly.DefinedTypes.ToList();
+
```
break between if/else and no `{}` 
i prefer the no `{}` but depends what the rest of the codebase uses

radumg commented on src/Engine/ProtoCore/FFI/CLRDLLModule.cs:
```diff
@@ -1230,9 +1397,16 @@ public FFIClassAttributes(Type type)
+
             foreach (var attr in attributes)
             {
+                //var type = Type.GetType(name);
```
commented code to delete

radumg commented on src/Engine/ProtoCore/FFI/CLRDLLModule.cs:
```diff
@@ -1426,9 +1616,9 @@ public string DefaultArgumentExpression
         }
 
-        public FFIParamAttributes(ParameterInfo parameter)
+        public FFIParamAttributes(ParameterInfo parameter, bool reflectionOnly)
```
suggest making this a separate ctor that just wraps the base so there's no breaking change to public API

```cs
public FFIParamAttributes(ParameterInfo parameter) // the original
public FFIParamAttributes(ParameterInfo parameter, bool reflectionOnly) // new one
private FFIParamAttributes(ParameterInfo parameter, bool reflectionOnly) // actual ctor
```

so the 2 public ones above would just be wrappers for the private one with `:this(param, reflectOnly)`

radumg commented on src/Engine/ProtoCore/FFI/FFIHandler.cs:
```diff
@@ -174,6 +176,26 @@ public static DLLModule GetModule(string dllModuleName)
+        public static CLRDLLModule GetModuleForInspection(Assembly assembly)
+        {
+            CLRDLLModule module;
+            var moduleName = assembly.GetName();
```
you never use `moduleName` directly below, it's always `moduleName.Name`, suggest doing that here in the assignment

radumg commented on src/Engine/ProtoCore/FFI/FFIHandler.cs:
```diff
@@ -174,6 +176,26 @@ public static DLLModule GetModule(string dllModuleName)
+            {
+                module = new CLRDLLModule(moduleName.Name, assembly);
+            }
+
```
delete break

radumg commented on src/Engine/ProtoCore/FFI/FFIHandler.cs:
```diff
@@ -174,6 +176,26 @@ public static DLLModule GetModule(string dllModuleName)
+                module = new CLRDLLModule(moduleName.Name, testDll);
+            }
+
+            Modules.Add(moduleName.Name + ".dll", module);
```
use interpolated string instead of concat (more efficient iirc?)

radumg commented on src/Engine/ProtoCore/Reflection/MemberInfoReflectionExtensions.cs:
```diff
@@ -0,0 +1,211 @@
+        }
+
+
+        internal static Attribute[] AttributesFromReflectionContext(this MemberInfo member)
```
suggest adding `Get` in front of these method names, the 4th one has it but not others - will be more descriptive to what method does

radumg commented on src/Engine/ProtoCore/Reflection/MemberInfoReflectionExtensions.cs:
```diff
@@ -0,0 +1,211 @@
+                    for (int i = 0; i < ca.ConstructorArguments.Count; i++)
+                    {
+                        var item = ca.ConstructorArguments[i];
+                        if (item.ArgumentType.IsArray)
```
suggest moving stuff inside these `if`s to private methods, this entire method is getting very long and hard to read

radumg commented on src/Engine/ProtoCore/Reflection/MemberInfoReflectionExtensions.cs:
```diff
@@ -0,0 +1,211 @@
+
+                    attributes.Add(a);
+                }
+                catch (Exception e)
```
remove exception if not used

radumg commented on src/Engine/ProtoCore/Reflection/MemberInfoReflectionExtensions.cs:
```diff
@@ -0,0 +1,211 @@
+
+
+                    dynamic ctorParam = null;
+                    if (paramValues.Count > 1)
+                    {
+                        var array = paramValues.ToArray<object>();
+                        if (TryConvertToTypeArray(array, out Array newArray))
+                        {
+                            ctorParam = newArray;
+                        }
+
+                        else
+                        {
+                            ctorParam = array;
+                        }
+                    }
+                    else
+                    {
+                        ctorParam = paramValues.FirstOrDefault();
+                    }  
```
add comment to what this block is doing, I don't understand it at first sight

radumg commented on src/Engine/ProtoCore/Reflection/MemberInfoReflectionExtensions.cs:
```diff
@@ -0,0 +1,211 @@
+
+        private static Attribute CreateInstance(System.Type type, object args)
+        {
+            var loadedAssembly = Assembly.LoadFrom(type.Assembly.Location);
```
can we check if assembly is already loaded before loading ? not sure if the assembly loading operation is idempotent 
cc @alvpickmans

radumg commented on src/Engine/ProtoCore/Reflection/MemberInfoReflectionExtensions.cs:
```diff
@@ -0,0 +1,211 @@
+            return attr;
+        }
+
+        private static object InstanceFromObject(this object obj, System.Type type)
```
suggest adding verb to method name `Make...`

radumg commented on src/NodeDocumentationMarkdownGenerator/AssemblyHandler.cs:
```diff
@@ -0,0 +1,375 @@
+        /// <param name="assemblyPaths">List of dll paths that should be scanned</param>
+        /// <param name="addtionalPathsToLoad">List of dll paths that should be added to the PathAssemblyResolver</param>
+        /// <returns></returns>
+        internal static List<MdFileInfo> ScanAssemblies(List<string> assemblyPaths, List<string> addtionalPathsToLoad, ILogger logger)
```
type in `addtionalPathsToLoad`

radumg commented on src/NodeDocumentationMarkdownGenerator/AssemblyHandler.cs:
```diff
@@ -0,0 +1,375 @@
+        /// <param name="assemblyPaths">List of dll paths that should be scanned</param>
+        /// <param name="addtionalPathsToLoad">List of dll paths that should be added to the PathAssemblyResolver</param>
+        /// <returns></returns>
+        internal static List<MdFileInfo> ScanAssemblies(List<string> assemblyPaths, List<string> addtionalPathsToLoad, ILogger logger)
```
make the `addtionalPathsToLoad` optional and then the ctor above this can just call this one, no more duplicated code

radumg replied:
or maybe we don't even need 2 of them ?

radumg commented on src/NodeDocumentationMarkdownGenerator/AssemblyHandler.cs:
```diff
@@ -0,0 +1,375 @@
+            using (mlc)
+            {
+                var dynamoCoreAss = resolver.Resolve(mlc, new AssemblyName("DynamoCore"));
+                var nodeModelType = dynamoCoreAss.GetType("Dynamo.Graph.Nodes.NodeModel");
```
`using Dynamo.Graph.Nodes;` is already added, do we not have access to this type directly ?
can do `typeof(NodeModel)` ?

radumg commented on src/NodeDocumentationMarkdownGenerator/AssemblyHandler.cs:
```diff
@@ -0,0 +1,375 @@
+                        Assembly asm = mlc.LoadFromAssemblyPath(path);
+                        AssemblyName name = asm.GetName();
+
+                        TryGetCustomizationFile(asm, out LibraryCustomization customization);
```
you usually use `TryGet` when you need to know if it succeeded or not.
we're not capturing this bool result in a `var` here and don't see any `if`s on `customization` further below, could this be a simpler `var customization = GetCustomizationForAssembly(asm)` method ?

radumg commented on src/NodeDocumentationMarkdownGenerator/AssemblyHandler.cs:
```diff
@@ -0,0 +1,375 @@
+                            catch { }
+                        }
+
+                        dllModule.ScanModule();
```
if it's not a `dll` or `exe`, should we still continue ?
that `dllModule` would be `null` and `dllModule.ScanModule()` would throw `NullException` below.

maybe we need to simply skip this loop iteration if it's not either of those 2 with a `continue;` ?

radumg commented on src/NodeDocumentationMarkdownGenerator/AssemblyHandler.cs:
```diff
@@ -0,0 +1,375 @@
+                                var externalLib = t.ClassNode.ExternLibName;
+
+                                // For some reason mscorelib sometimes gets pass to here, so filtering it away.
+                                if (t.CLRType.Assembly.GetName().Name == typeof(object).Assembly.GetName().Name ||
```
cache `typeof(object).Assembly.GetName().Name` into a var outside loop so we don't call it on every nested loop iteration

radumg commented on src/NodeDocumentationMarkdownGenerator/AssemblyHandler.cs:
```diff
@@ -0,0 +1,375 @@
+
+                                var functionNodes = t.ClassNode.Procedures
+                                    .Where(c => c is FunctionDefinitionNode)
+                                    .Cast<FunctionDefinitionNode>()
```
`OfType<>`

radumg commented on src/NodeDocumentationMarkdownGenerator/AssemblyHandler.cs:
```diff
@@ -0,0 +1,375 @@
+
+                                var ctorNodes = t.ClassNode.Procedures
+                                    .Where(c => c.Kind == AstKind.Constructor)
+                                    .Cast<ConstructorDefinitionNode>()
```
`OfType<>`

radumg commented on src/NodeDocumentationMarkdownGenerator/AssemblyHandler.cs:
```diff
@@ -0,0 +1,375 @@
+                                    .Cast<AssociativeNode>();
+
+                                var associativeNodes = ctorNodes.ToList();
+                                if (functionNodes.Any()) associativeNodes.AddRange(functionNodes);
```
you could do a `Union` on these 2 lists or do `var associativeNodes = new List<>(ctorNodes, functionNodes)` to auto-concat them?

radumg commented on src/NodeDocumentationMarkdownGenerator/AssemblyHandler.cs:
```diff
@@ -0,0 +1,375 @@
+                    }
+                }
+
+                AddZeroTouchNodesToSearch(functionGroups.Values.SelectMany(x => x.Values.Select(g => g)), ref nodeSearchModel);
```
could use `Flatten` here maybe ? instead of the `SelectMany( Select() ) `?

radumg commented on src/NodeDocumentationMarkdownGenerator/AssemblyHandler.cs:
```diff
@@ -0,0 +1,375 @@
+                        arg.NameNode.Name :
+                        arg.NameNode.Kind == AstKind.BinaryExpression ?
+                            ((BinaryExpressionNode)arg.NameNode).LeftNode.Name :
+                            string.Empty;
```
🤯

radumg replied:
lost me here in the nested-ness

radumg commented on src/NodeDocumentationMarkdownGenerator/Commands/FromDirectoryCommand.cs:
```diff
@@ -0,0 +1,65 @@
+
+        private List<MdFileInfo> ScanFolderForCustomNodes(string inputFolderPath, SearchOption searchOption)
+        {
+            var allDyfs = Directory.GetFiles(inputFolderPath, "*.dyf", searchOption).Select(x => new FileInfo(x)).ToList();
```
extract `*.dyf`, `*.dll` and others into constants ? used in a few places (other files in project too)

radumg commented on src/NodeDocumentationMarkdownGenerator/Commands/FromDirectoryCommand.cs:
```diff
@@ -0,0 +1,65 @@
+
+            return AssemblyHandler.
+                ScanAssemblies(allDlls.
+                Select(x => x.FullName).ToList(), logger);
```
this select is inside the `ScanAssemblies`, so should be on same line or indented further

radumg commented on src/NodeDocumentationMarkdownGenerator/MarkdownHandler.cs:
```diff
@@ -0,0 +1,298 @@
+                var mainDirectory = new FileInfo(dictionaryPath).Directory;
+                examplesDirectory = Path.Combine(mainDirectory.FullName, "EXAMPLES");
+
+                if (compressImages)
+                {
+                    optimizer = new ImageOptimizer();
+                    optimizer.OptimalCompression = true;
+                }
```
should the image compression only happen if the dictionary path is provided ? 
ie, should we not also optimise images for ZT authors if they set it to true ?

SHKnudsen replied:
For now i removed the option for a dictionary path for package authors. The sturcuture of the dictionary json is.... well, not great, so if that sturcture is to be used by 3rd party devs it probably needs a little work

radumg commented on src/NodeDocumentationMarkdownGenerator/MarkdownHandler.cs:
```diff
@@ -0,0 +1,298 @@
+
+                if (File.Exists(filePath))
+                {
+                    var t = fileName;
```
`t` would be scoped to this `if` block and hence this line/block does nothing ?

radumg commented on src/NodeDocumentationMarkdownGenerator/MarkdownHandler.cs:
```diff
@@ -0,0 +1,298 @@
+            var content = new StringBuilder();
+            content.AppendLine($"## {nodeName} - Documentation");
+            content.AppendLine($"This documentation file is auto generated by {Assembly.GetExecutingAssembly().GetName()}.");
+            content.AppendLine($"\nFor more information about adding documentation to nodes see https://github.com/DynamoDS/Dynamo/wiki/Create-and-Add-Custom-Documentation-to-Nodes");
```
extract URL to constant or `*.resx` file

radumg replied:
actually, might be good to also do that for these texts so they're translatable

radumg commented on src/NodeDocumentationMarkdownGenerator/MarkdownHandler.cs:
```diff
@@ -0,0 +1,298 @@
+
+        }
+
+        private static DynamoDictionaryEntry GetMatchigDictionaryEntry(List<DynamoDictionaryEntry> dictEntrys, MdFileInfo info, LayoutSpecification spec)
```
typo in method name GetMatchi`n`gDictionaryEntry

radumg commented on src/NodeDocumentationMarkdownGenerator/MarkdownHandler.cs:
```diff
@@ -0,0 +1,298 @@
+            return content.ToString();
+        }
+
+        private static string GetContentFormDictionaryEntry(DynamoDictionaryEntry entry, string examplesDirectory, ImageOptimizer optimizer, FileInfo fileInfo, ILogger logger)
```
is the method name intended to be `From` and not `Form` ?

radumg commented on src/NodeDocumentationMarkdownGenerator/MdFileInfo.cs:
```diff
@@ -0,0 +1,121 @@
+
+        public string FullCategory { get; }
+
+        public string QualifiedFileName { get; }
```
add comments to all props and class itself - others new to this codebase should have instructions on what this represents, etc

radumg replied:
for example, not sure what this `QualifiedFileName` is  ?

radumg commented on src/NodeDocumentationMarkdownGenerator/MdFileInfo.cs:
```diff
@@ -0,0 +1,121 @@
+        /// <summary>
+        /// Name of external library this node belongs to
+        /// </summary>
+        public string ExternalLib { get; }
```
a few extra chars makes it clearer
```suggestion
        public string ExternalLibraryName { get; }
```

radumg commented on src/NodeDocumentationMarkdownGenerator/Program.cs:
```diff
@@ -0,0 +1,40 @@
+            var header = @"
+ _______                                                        
+/       \                                                       
+$$$$$$$  | __    __  _______    ______   _____  ____    ______  
```
could do a version that also has the logo ? 😁

radumg commented on src/NodeDocumentationMarkdownGenerator/Properties/launchSettings.json:
```diff
@@ -0,0 +1,8 @@
+  "profiles": {
+    "NodeDocumentationMarkdownGenerator": {
+      "commandName": "Project",
+      "commandLineArgs": "fromdirectory -i \"C:\\Users\\SylvesterKnudsen\\Documents\\GitHub\\DynamoRevit\\bin\\AnyCPU\\Release\\Revit\" -f RevitNodes.dll -o \"C:\\Users\\SylvesterKnudsen\\Desktop\\TestMdOutputRevitNodes - Compressed\" -d \"C:\\Users\\SylvesterKnudsen\\Documents\\GitHub\\DynamoDictionary\\public\\data\\Dynamo_Nodes_Revit.json\" -l \"C:\\Users\\SylvesterKnudsen\\Desktop\\Logs\" -x  ...
```
file should probs not be committed to repo ?
or have a non user-specific path if that's possible ?

SHKnudsen committed:
add support for .ds files

SHKnudsen committed:
comments + move to tools folder

SHKnudsen committed:
add reference to RestSharp - need to create PackageUploadRequestBody

SHKnudsen committed:
Update FromPackageFolderCommand.cs

SHKnudsen committed:
Merge remote-tracking branch 'upstream/master' into Node-Markdown-generator-tool

SHKnudsen committed:
🧹

SHKnudsen committed:
Remove logger

SHKnudsen committed:
remove unused using

SHKnudsen committed:
Update FromDirectoryOptions.cs

SHKnudsen committed:
Update AssemblyHandler.cs

SHKnudsen committed:
Change project type + add test project

SHKnudsen committed:
remove old files

SHKnudsen committed:
use flag to specify ReflectionContext

SHKnudsen committed:
Merge remote-tracking branch 'upstream/master' into Node-Markdown-generator-tool

SHKnudsen committed:
few comment updates

SHKnudsen committed:
Update DocumentationBrowserViewExtensionTests.cs

SHKnudsen committed:
ds file fixes

SHKnudsen committed:
fix add isReflection flag FFIMethodAtribute

SHKnudsen committed:
First pass at comments

SHKnudsen committed:
Update MarkdownHandler.cs

SHKnudsen committed:
2nd pass on comments

SHKnudsen committed:
Update MarkdownHandler.cs

SHKnudsen committed:
help text updates

SHKnudsen committed:
update licence

SHKnudsen committed:
comments

SHKnudsen committed:
add reflection context object type check to CLRDLLModule

SHKnudsen committed:
Use IconName to get the type FullName from a NodeModelSearchElement + other clean ups

SHKnudsen committed:
get baseType with overloads

SHKnudsen committed:
Load dynamo dlls from main bin folder + fix overload lookup

SHKnudsen committed:
Merge remote-tracking branch 'upstream/master' into Node-Markdown-generator-tool

SHKnudsen committed:
Update DocumentationBrowserViewExtensionTests.cs

mjkkirschner committed:
fix issue with protogeometry format load error
put clr into reflection mode when importing DS files
fix classname for generated ds docs

mjkkirschner committed:
remove unused

SHKnudsen committed:
Merge pull request #25 from mjkkirschner/nmgt_mjk

fix issue with protogeometry format load error

SHKnudsen committed:
remove test project

mjkkirschner committed:
Merge branch 'master' into Node-Markdown-generator-tool

SHKnudsen committed:
comment cleanups

mjkkirschner committed:
merge license (#31)

Co-authored-by: michael kirschner <michael.kirschner@autodesk.com>

mjkkirschner committed:
move LayoutSpecification types to DynamoCore with type forward (#30)

Co-authored-by: michael kirschner <michael.kirschner@autodesk.com>

mjkkirschner committed:
add new flag -g to compress gifs
add gif compression by quantization to 32k colors
in debug leave console open until readline for easy log reading

mjkkirschner committed:
typo

mjkkirschner committed:
try to resolve merge conflicts with dynamo master (#33)

* Dyn 3817 exit tour window popup (#11918)

* DYN-3817-ExitTourWindow Popup

I added the UI for the Exit Tour Window, but still having some issues in the layout

* DYN-3817-ExitTourWindow-Popup

I created the view for the Exit Tour Window and also the functionality of showing the ExitTour ( this is shown only when one of the steps is closed so the tour is not finished - this doesn't apply for the Survey popup).
A resource string was added and is related to the ExitTour content so it can be localized later.

* DYN-3817-ExitTourWindow-Popup

Removing code not needed in DynamoView.xaml.cs

* DYN-3817-ExitTourWindow-Popup Code Review

The ExitTourWindow was renamed to RealTimeInfoWindow.
I reorganized the RealTimeInfoWindow so doesn't depend of a Step (then it can be used in a more generic way).
I created the properties Width, Height and TextContent inside the RealTimeInfoWindow  so they can be set up from outside. The TextContent property contains the text that will be shown in the popup and it can be updated on runtime.

* DYN-3817-ExitTourWindow-Popup Code Issues

Fixing issues produced after merging the latest changes in master branch

* DYN-3817-ExitTourWindow-Popup Code Review 2

The offset hardcoded values were changed to const double, also I reorganized the way the RealTimeInfoWindow object is created.
Also I added some comments in the xaml file

* Dyn-3640 - Recover execution after clearing cyclic dependency from graph (#11896)

* cleanup

* minor cleanup

* clear cycles from dependency graph nodes for graph runs

* cleanup

* update failing test

* clear node warnings for cleared cycles

* do not include cyclic graph node as a redefined node of another graph node

* add unit tests

* fix one more case

* review comments

* add logging to intermittent pkgloader tests (#11925)

* update tests

* update xml summary
add some logging to failing tests

Co-authored-by: michael kirschner <michael.kirschner@autodesk.com>

* pkg load test log typo (#11926)

* update tests

* update xml summary
add some logging to failing tests

* typo

Co-authored-by: michael kirschner <michael.kirschner@autodesk.com>

* Update UI (#11932)

* Update searchMethod invocation in LibraryMSWebBrowserExtension (#11931)

* Update searchMethod invocation 

Update searchMethod invocation  to use correct number of parameters

* Update SearchDictionary.cs

* Add adsk fonts to extern folder with license terms. (#11890)

* small adjustment

* align titles

* add names to buttons for ui automation tests

* missing button name

* new font folder for use in Dynamo only

* Update readme.md

* add font links to about box

Co-authored-by: michael kirschner <michael.kirschner@autodesk.com>

* check for null (#11930)

* Prevent port measure crash (#11935)

* CRASH: Delete node with right-click at the same time (#11927)

* null checks

* Add unit test

* Fix For Loading Json File (#11934)

* Fix For Loading Json File

During the testing of the task https://jira.autodesk.com/browse/DYN-3817 Aabishkar reported that when trying to open the "Get Started" tour a message appear saying Could not find a part of the path 'C:\Windows\system32\UI\GuidedTour\dynamo_guides.json'
Then I modified the code that is loading the json,file, in the past was a relative path and now we are using an absolute path.

* Fix For Loading Json File

Instead to have the json file path in the GuidesManager constructor I created two public properties that will contains the assembly path execution and the JSON files guides path, so we can verify in the unit testing that this paths exists and if the Guide will be created correctly.
Also I added a unit test that will validate that the json guides file exists and that the Guides can be created.

* Handle crash due to save path being too long

Co-authored-by: Roberto T <61755417+RobertGlobant20@users.noreply.github.com>
Co-authored-by: aparajit-pratap <aparajit.pratap@autodesk.com>
Co-authored-by: michael kirschner <michael.kirschner@autodesk.com>
Co-authored-by: Aaron (Qilong) <173288704@qq.com>
Co-authored-by: Ashish Aggarwal <ashish.aggarwal@autodesk.com>

mjkkirschner committed:
Merge branch 'master' into Node-Markdown-generator-tool

SHKnudsen committed:
Merge pull request #32 from mjkkirschner/nmdg-imagecompgif

add new flag -g to compress gifs

SHKnudsen committed:
add dyf folder exists check
